File: football-api/app/core/job_metrics.py
"""Heartbeat Prometheus dos jobs do APScheduler (PRD 049, item 4 — observabilidade).

Métricas de *liveness*, não de resultado de negócio: `record_job_success` marca que
a execução completou, mesmo que nenhuma partida tenha sido criada/notificada.
Alimentam os alertas Grafana "Scheduler" (monitoring/grafana/provisioning/alerting/rules.yml),
que detectam o scheduler mudo após saturação do pooler Supabase (EMAXCONNSESSION).

O label `pid` distingue os 2 workers uvicorn de produção: cada scrape do Prometheus
cai num worker aleatório e, sem o label, o counter alternaria entre valores absolutos
de processos diferentes (resets perpétuos aos olhos do `increase()`).
"""
import os

from prometheus_client import Counter, Gauge

JOB_RECURRENCE = "recurrence"
JOB_STATUS_SYNC = "status_sync"
JOB_VOTE_REMINDER = "vote_reminder"
ALL_JOBS = (JOB_RECURRENCE, JOB_STATUS_SYNC, JOB_VOTE_REMINDER)

JOB_LAST_SUCCESS = Gauge(
    "scheduler_job_last_success_timestamp_seconds",
    "Unix timestamp do último sucesso de cada job do APScheduler",
    ["job_name", "pid"],
)

# Exposto como scheduler_job_failures_total (o client acrescenta o sufixo _total)
JOB_FAILURES = Counter(
    "scheduler_job_failures",
    "Total de falhas por job do APScheduler",
    ["job_name", "pid"],
)


def record_job_success(job_name: str) -> None:
    JOB_LAST_SUCCESS.labels(job_name=job_name, pid=str(os.getpid())).set_to_current_time()


def record_job_failure(job_name: str) -> None:
    JOB_FAILURES.labels(job_name=job_name, pid=str(os.getpid())).inc()


def init_job_metrics() -> None:
    """Cria as séries no startup (gauge = agora, counter = 0).

    Sem isto, o alerta de staleness operaria sobre métrica ausente até a primeira
    execução de cada job, e um counter nascido em N esconderia as N primeiras
    falhas do `increase()`.
    """
    for name in ALL_JOBS:
        record_job_success(name)
        JOB_FAILURES.labels(job_name=name, pid=str(os.getpid()))

File: football-api-go/internal/services/job_metrics.go
package services

import (
	"github.com/prometheus/client_golang/prometheus"
	"github.com/prometheus/client_golang/prometheus/promauto"
)

// Heartbeat Prometheus dos jobs do scheduler (PRD 049 §4).
// Espelha football-api/app/core/job_metrics.py — mesmos nomes de métrica e de
// job, para que as queries e alertas escritos para a v1 funcionem sem adaptação.
// Diferença deliberada: sem o label `pid` (a v1 roda 2 workers uvicorn; aqui o
// processo é único e as queries usam max()/sum(), que seguem válidas).

const (
	JobRecurrence   = "recurrence"
	JobStatusSync   = "status_sync"
	JobVoteReminder = "vote_reminder"
)

var allJobs = []string{JobRecurrence, JobStatusSync, JobVoteReminder}

var (
	jobLastSuccess = promauto.NewGaugeVec(prometheus.GaugeOpts{
		Name: "scheduler_job_last_success_timestamp_seconds",
		Help: "Unix timestamp do último sucesso de cada job do scheduler.",
	}, []string{"job_name"})

	// Exposta como scheduler_job_failures_total (o client adiciona _total no
	// client_python da v1; aqui o sufixo é explícito para o nome final coincidir).
	jobFailures = promauto.NewCounterVec(prometheus.CounterOpts{
		Name: "scheduler_job_failures_total",
		Help: "Total de execuções com falha de cada job do scheduler.",
	}, []string{"job_name"})
)

// RecordJobSuccess marca o heartbeat de liveness do job — deve ser chamada em
// todo término sem erro, mesmo quando o job não teve nada a fazer.
func RecordJobSuccess(jobName string) {
	jobLastSuccess.WithLabelValues(jobName).SetToCurrentTime()
}

// RecordJobFailure incrementa o contador de falhas do job.
func RecordJobFailure(jobName string) {
	jobFailures.WithLabelValues(jobName).Inc()
}

// InitJobMetrics inicializa as séries dos jobs no boot (gauge = agora,
// counter = 0), evitando staleness no `time() - max(...)` dos alertas e
// increase() sem série base logo após um restart.
func InitJobMetrics() {
	for _, job := range allJobs {
		jobLastSuccess.WithLabelValues(job).SetToCurrentTime()
		jobFailures.WithLabelValues(job).Add(0)
	}
}
